src/webgpu/util/fp_interval.ts

```diff
@@ -0,0 +1,238 @@
+    this.n = Math.abs(n);
+  }
+
+  public singular(x: number): FPInterval {
+    assert(!Number.isNaN(x), `absolute not defined for NaN`);
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  public range(i: FPInterval): FPInterval {
+    if (i.begin === i.end) {
+      return this.singular(i.begin);
+    }
+    return FPInterval.span(this.singular(i.begin), this.singular(i.end));
+  }
+
+  private impl(x: number): FPInterval {
+    return new FPInterval(x - this.n, x + this.n);
+  }
+}
+
+class CorrectlyRoundedFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  public constructor() {}
+
+  public singular(x: number): FPInterval {
+    assert(!Number.isNaN(x), `correctlyRounded not defined for NaN`);
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  public range(i: FPInterval): FPInterval {
+    if (i.begin === i.end) {
+      return this.singular(i.begin);
+    }
+    return FPInterval.span(this.singular(i.begin), this.singular(i.end));
+  }
+
+  private impl(x: number): FPInterval {
+    if (x === Number.POSITIVE_INFINITY || x > kValue.f32.positive.max) {
+      return new FPInterval(kValue.f32.positive.max, Number.POSITIVE_INFINITY);
+    }
+
+    if (x === Number.NEGATIVE_INFINITY || x < kValue.f32.negative.min) {
+      return new FPInterval(Number.NEGATIVE_INFINITY, kValue.f32.negative.min);
+    }
+
+    const x_32 = new Float32Array([x])[0];
+    const converted: number = x_32;
+    if (x === converted) {
+      // x is precisely expressible as a f32, so correctly rounding degrades to exactly matching
+      return new FPInterval(x, x);
+    }
+
+    if (converted > x) {
+      // x_32 rounded towards +inf, so is after x
+      const otherside = nextAfter(x_32, false, false).value as number;
+      return new FPInterval(otherside, converted);
+    } else {
+      // x_32 rounded towards -inf, so is before x
+      const otherside = nextAfter(x_32, false, false).value as number;
+      return new FPInterval(converted, otherside);
+    }
+  }
+}
+
+class ULPFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  protected readonly n: number;
+
+  public constructor(n: number) {
+    this.n = Math.abs(n);
+  }
+
+  public singular(x: number): FPInterval {
+    assert(!Number.isNaN(x), `ULP not defined for NaN`);
+    return FPInterval.span(this.impl(x), this.impl(flushSubnormalNumber(x)));
+  }
+
+  public range(i: FPInterval): FPInterval {
+    if (i.begin === i.end) {
+      return this.singular(i.begin);
+    }
+    return FPInterval.span(this.singular(i.begin), this.singular(i.end));
+  }
+
+  private impl(x: number): FPInterval {
+    const ulp_flush = oneULP(x, true);
+    const ulp_noflush = oneULP(x, false);
+    const ulp = Math.max(ulp_flush, ulp_noflush);
+
+    return new FPInterval(x - this.n * ulp, x + this.n * ulp);
+  }
+}
+
+export class CosFPIntervalBuilder implements FPUnaryIntervalBuilder {
+  private readonly builder;
```

Comment:
I am choosing to use composition here over inheritance, so that Builders that don't need a range() implementation don't inherit one from their parent